costmap_2d/src/footprint.cpp

```diff
@@ -66,6 +66,39 @@ void calculateMinAndMaxDistances(const std::vector<geometry_msgs::Point>& footpr
   max_dist = std::max(max_dist, std::max(vertex_dist, edge_dist));
 }
 
+double minSweepingAreaOrientation(const std::vector<geometry_msgs::Point>& footprint)
+{
+  double min_dist = std::numeric_limits<double>::max();
+  std::vector<geometry_msgs::Point> closest_edge;
+
+  if (footprint.size() <= 2)
+  {
+    return NAN;
+  }
+
+  // check the distance from the robot center point to each footprint edged and keep the closest one
+  for (unsigned int i = 0; i < footprint.size() - 1; ++i)
+  {
+    double edge_dist = distanceToLine(0, 0, footprint[i].x, footprint[i].y, footprint[i + 1].x, footprint[i + 1].y);
+    if (edge_dist < min_dist)
```

Comment:
Q: I don't understand how distance to the center will give you the smallest sweeping area. 
A rectangle like the one below will give you 90 degrees as output (because the left side is the closest to the center). But 90 degrees would sweep the largest area
![rectangle](https://user-images.githubusercontent.com/6097267/182052288-96e3d097-240e-4db5-9936-ac4d752254cb.png)

Reply:
if +x is the horizontal, rotating the rectangle 90 deg will minimize the sweeping area;  so 90 deg is the right answer

Reply:
but for a "kite" shape it doesn't work, because it should be 90 deg too, and it is not:
![image](https://user-images.githubusercontent.com/6097267/182140578-e1876e5e-ed6b-48aa-8fab-628a11c43bc9.png)

Reply:
the answer won't be 90, true
but it will be correct anyway: will sweep the same area as with 90

in any case that fp violates the 2nd assumption I added to the function (true, the assumptions are a bit ad-hoc,,, but are true for every polygonal robot I know

Reply:
wdym it will sweep the same area? the angle would be ~78, and if you rotate the swept area is way larger:
![image](https://user-images.githubusercontent.com/6097267/182163299-e22f19e0-830b-4c4c-9824-6f42cebb676b.png)

hum...need to think about those two assumptions, but it might be easier to implement the generic case

Reply:
hum, true. But then the function should check the violation. You don't need to loop all edges if you assume the closest one is parallel to the x-axis